MIP/mip-61/README.md

```diff
@@ -0,0 +1,297 @@
+
+```javascript
+// Relayer Local Structure
+{
+    blocks: [                          // Array of blocks being tracked
+        {
+            blockHeight: Number,       // Height of the block
+            status: String,            // Status of the block (e.g., "processed", "completed")
+            events: [                  // Array of events in the block
+                {
+                    eventUID: String,  // Unique identifier for the event
+                    nonce: Number,     // Matching nonce from the source chain
+                    status: String     // Status of the event (e.g., "pending", "completed")
+                }
+            ]
+        }
+    ]
+}
+```
+
+
+### EVENT_PROCESSING
+
+The relayer may send a message `complete_transfer` to the target chain for an event that has the status `transfer_pending`. This status exists during the time between the reception of the `initiate_transfer` event from the source chain and the success of the `complete_transfer` transaction on the target chain. (However, it is possible that a `complete_transfer` messages is not yet finalized on the target chain, in which case the relayer DOES NOT send a new `complete_transfer` message, but waits instead.)
+
+The Native Bridge protocol MUST implement the assignment of an incrementing `nonce` on the source chain. Since there are two directions there MUST be two types of counters (`nonces`) - one for each transfer direction. The `nonce` is used to order the transfers.
+
+For each `initiate_transfer` event do the following:
+
+![event_processing](event_processing.png)
+
+```javascript
+EVENT_PROCESSING:
+// Input: initiate_transfer_event from source chain
+// Input: target_chain_state - current state of the target chain
+SET `nonce` = `initiate_transfer_event.nonce`
+SET `transfer_uid` = `initiate_transfer_event.transfer_uid`
+
+// Check if the `nonce` is already recorded
+IF `nonce` is recorded THEN:
+    IF `nonce.status_transfer` is `transfer_completed` THEN:
+        RETURN.
+ELSE: 
+    CREATE new `nonce` entry locally
+
+QUERY target chain contract for `nonce` // this includes non-final state!
+        
+// If `complete_transfer` transaction was sent previously to target chain
+IF `nonce` is found THEN:
+    IF `transfer_uid` does not match THEN:
+        REPORT error and EXIT
+
+    IF `nonce` transfer is final THEN:
+        SET `nonce.status_transfer` to `transfer_completed`
+    ELSE:
+        // the status should be `transfer_pending`
+        IF `nonce.status_transfer` is NOT `transfer_pending` THEN:
+            LOG "error: nonce.status_transfer is not transfer_pending"
+        SET `nonce.status_transfer` to `transfer_pending`
+        SET timeout for relayer to re-check later
+// If no `complete_transfer` transaction was sent 
+ELSE:
+    SEND `complete_transfer` transaction to target chain
+    SET `nonce.status_transfer` to `transfer_pending`
+    SET timeout for relayer to re-check later
+```
+
```

Comment:
There's a missing part in this algo. For normal processing, when the nonce doesn't exist, and we send a complete transfer call, the status of the nonce will stay forever in transfer_pending. We need to monitor the target chain for transfer_completed event in finalized state to change this status.
Another point, pending transfer are always replaying. In our algo we replay it only if the sent complete transfer Tx failed.